File: LeetCode/700二叉搜索树中的搜索.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


struct TreeNode* searchBST(struct TreeNode* root, int val){
    if (root == NULL) return NULL;
    if (root->val == val) return root;
    return searchBST(val > root->val ? root->right : root->left, val);
}


/**
 * C++
 */
class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        if (root == nullptr) return root;
        if (root->val < val) {
            return searchBST(root->right, val);
        } else if (root->val > val) {
            return searchBST(root->left, val);
        }
        return root;
    }
};

File: LeetCode/700二叉搜索树中的搜索.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func searchBST(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return root
	}
	for root != nil {
		if root.Val < val {
			root = root.Right
		} else if root.Val > val {
			root = root.Left
		} else {
			return root
		}
	}
	return nil
}